numbat/src/vm.rs

```diff
@@ -946,6 +978,34 @@ impl Vm {
                         self.stack.push(return_value);
                     }
                 }
+                Op::BuildStruct => {
+                    let info_idx = self.read_u16();
+                    let (_, struct_info) = self
+                        .struct_infos
+                        .get_index(info_idx as usize)
+                        .expect("Missing struct metadata");
+                    let struct_info = Arc::clone(struct_info);
+                    let num_args = self.read_u16();
+
+                    let mut content = Vec::with_capacity(num_args as usize);
+
+                    for _ in 0..num_args {
+                        content.push(self.pop());
+                    }
+
+                    self.stack.push(Value::Struct(struct_info, content));
+                }
+                Op::DestructureStruct => {
+                    let field_idx = self.read_u16();
+                    let content = self.pop();
+
+                    let Value::Struct(_meta, mut content) = content else {
+                        panic!("tried to destructure something which was not a struct (the typechecker should prevent this)");
+                    };
```

Comment:
For other `Value` types, we have `.unsafe_as_xxx()` methods. Let's maybe follow the same pattern here (see `impl Value` in `value.rs`)